File: Shortest Path/Floyd-Warshall/2610번 - 회의준비/2610.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

#define MAX 101
#define INF 9876

int n, m;
vector<int> person[MAX], group, delegate;
bool vis[MAX];
int floyd[MAX][MAX];

void input() {
    cin >> n >> m;
    fill_n(&floyd[0][0], MAX * MAX, INF);
    for(int i = 0, a, b; i < m; i++) {
        cin >> a >> b;
        person[a].push_back(b);
        person[b].push_back(a);
        floyd[a][b] = 1;
        floyd[b][a] = 1;
    }
}

void dfs(int node) {
    vis[node] = true;
    group.push_back(node);
    for(int i = 0; i < person[node].size(); i++) {
        int next_node = person[node][i];
        if(!vis[next_node]) dfs(next_node);
    }
}

int head(vector<int> people) {
    int dist_val = INF;
    int headpoint = -1;
    
    for(int i = 0; i < people.size(); i++) {
        int cur = people[i];
        int max_val = -1;
        for(int j = 1; j <= n; j++) {
            if(floyd[cur][j] == INF) continue;
            max_val = max(max_val, floyd[cur][j]);
        }
        
        // new head
        if(dist_val > max_val) {
            headpoint = cur;
            dist_val = max_val;
        }
    }
    return headpoint;
}

void solve() {
    for(int k = 1; k <= n; k++) {
        for(int i = 1; i <= n; i++) {
            for(int j = 1; j <= n; j++) {
                if(i == j) floyd[i][j] = 0;
                else floyd[i][j] = min(floyd[i][j], floyd[i][k] + floyd[k][j]);
            }
        }
    }
    
    for(int i = 1; i <= n; i++) {
        if(!vis[i]) {
            group.clear();
            dfs(i);
            int headpoint = head(group);
            if(headpoint != -1) delegate.push_back(headpoint);
        }
    }
    
    sort(delegate.begin(), delegate.end());
    cout << delegate.size() << '\n';
    for(auto &it : delegate) cout << it << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Shortest Path/Floyd-Warshall/2610번 - 회의준비/2610.py
import sys
from collections import deque

input = sys.stdin.readline

MAX = 101
INF = 10**2

n = int(input())
m = int(input())
relation = list([] for _ in range(n + 1))
vis = [False] * (n + 1)
floyd = [[INF] * (n + 1) for _ in range(n + 1)]
delegate = []
for _ in range(m):
    a, b = map(int, input().split())
    relation[a].append(b)
    relation[b].append(a)
    floyd[a][b] = floyd[b][a] = 1

def bfs(node, arr):
    Q = deque()
    Q.append(node)
    vis[node] = True

    while Q:
        cur = Q.pop()
        arr.append(cur)
        for nextNode in relation[cur]:
            if not vis[nextNode]:
                Q.append(nextNode)
                vis[nextNode] = True

    return arr
    
def groupPoint(group):
    distVal = INF
    curPoint = -1
    
    for person in group:
        maxVal = -1
        for i in range(n + 1):
            if floyd[person][i] == INF:
                continue
            maxVal = max(maxVal, floyd[person][i])
        
        if distVal > maxVal:
            distVal = maxVal
            curPoint = person
    
    return curPoint

def solve():
    for k in range(1, n + 1):
        for i in range(1, n + 1):
            for j in range(1, n + 1):
                if i == j:
                    floyd[i][j] = 0
                else:
                    floyd[i][j] = min(floyd[i][j], floyd[i][k] + floyd[k][j])
    
    for i in range(1, n + 1):
        if not vis[i]:
            group = bfs(i, [])
            delegate.append(groupPoint(group))
    
    sortedDelegate = sorted(delegate)
    sortedDelegate = [len(sortedDelegate)] + sortedDelegate
    print('\n'.join(map(str, sortedDelegate)))

if __name__ == '__main__':
    solve()